### ISSUE TITLE
显存占用过大

### ISSUE THREAD

**Issue by Feynman1999** [MegEngine/MegEngine#81]
## 环境
1.系统环境：ubantu18.04
2.MegEngine版本：0.6
3.python版本：3.7

## 复现步骤
1.实现了超分辨的经典网络DBPN
2.发现显存占用过大，相比torch在3倍以上

## 请提供关键的代码片段便于追查问题
```
import megengine.module as M
from megengine.module.conv import Conv2d, ConvTranspose2d
import megengine.functional as F

class UPU(M.Module):
    """up-projection units"""

    def __init__(self,
                 num_channels,
                 filter_size,
                 stride,
                 padding):
        super(UPU, self).__init__()
        self.deconv1 = M.Sequential(
            ConvTranspose2d(num_channels, num_channels, filter_size, stride, padding),
            M.PReLU(num_parameters=1, init=0.25)
        )
        self.conv1 = M.Sequential(
            Conv2d(num_channels, num_channels, filter_size, stride, padding),
            M.PReLU(num_parameters=1, init=0.25)
        )
        self.deconv2 = M.Sequential(
            ConvTranspose2d(num_channels, num_channels, filter_size, stride, padding),
            M.PReLU(num_parameters=1, init=0.25)
        )

    def forward(self, x1):
        x2 = self.deconv1(x1)
        x3 = self.conv1(x2)
        x3 = x3 - x1
        x4 = self.deconv2(x3)
        return x4 + x2


class DPU(M.Module):
    """
        down-projection units
    """
    def __init__(self, num_channels, filter_size, stride, padding):
        super(DPU, self).__init__()
        self.conv1 = M.Sequential(
            Conv2d(num_channels, num_channels, filter_size, stride, padding),
            M.PReLU(num_parameters=1, init=0.25)
        )
        self.deconv1 = M.Sequential(
            ConvTranspose2d(num_channels, num_channels, filter_size, stride, padding),
            M.PReLU(num_parameters=1, init=0.25)
        )
        self.conv2 = M.Sequential(
            Conv2d(num_channels, num_channels, filter_size, stride, padding),
            M.PReLU(num_parameters=1, init=0.25)
        )

    def forward(self, x1):
        x2 = self.conv1(x1)
        x3 = self.deconv1(x2)
        x3 = x3 - x1
        x4 = self.conv2(x3)
        return x4 + x2


class DBPN(M.Module):
    """DBPN network structure.

    Paper:
    Ref repo:

    Args:
    """

    def __init__(self,
                 in_channels=3,
                 out_channels=3,
                 n_0=256,
                 n_R=64,
                 iterations_num=10,
                 upscale_factor=4):
        super(DBPN, self).__init__()

        filter_size = upscale_factor + 4
        stride = upscale_factor
        padding = 2
        self.iterations_num = iterations_num
        self.conv1 = M.Sequential(
            Conv2d(in_channels, n_0, 3, 1, 1),
            M.PReLU(num_parameters=1, init=0.25)
        )
        self.conv2 = M.Sequential(
            Conv2d(n_0, n_R, 1, 1, 0),
            M.PReLU(num_parameters=1, init=0.25)
        )
        self.UPU = UPU(n_R, filter_size, stride, padding)
        self.DPU = DPU(n_R, filter_size, stride, padding)
        self.conv3 = M.Sequential(
            Conv2d(n_R*iterations_num, out_channels, 3, 1, 1),
            M.PReLU(num_parameters=1, init=0.25)
        )

    def forward(self, x):
        """Forward function.

        Args:
            x (Tensor): Input tensor with shape (n, c, h, w).

        Returns:
            Tensor: Forward results.
        """
        H_list = []
        x = self.conv1(x)
        x = self.conv2(x)

        for _ in range(self.iterations_num-1):
            H = self.UPU(x)
            H_list.append(H)
            x = self.DPU(H)

        H_list.append(self.UPU(x))
        x = F.concat(H_list, axis=1)
        x = self.conv3(x)
        return x

    def init_weights(self, pretrained=None, strict=True):
        pass

```

## 请提供完整的日志及报错信息
可以正常运行，并且精度正常，就是显存占用大，batchsize调大后会OOM

**Comment by ChaiMind** [comment]
@FateScript Please take a look.

**Comment by xxr3376** [comment]
@Feynman1999 请提供一下训练部分的代码，谢谢

**Comment by Feynman1999** [comment]
## 训练部分代码
```
@jit.trace(symbolic=True)
def train_generator_batch(image, label, *, opt, netG, netloss):
    netG.train()
    output = netG(image)
    loss = netloss(output, label)
    opt.backward(loss)
    if dist.is_distributed():
        # do all reduce mean
        pass
    return loss


@jit.trace(symbolic=True)
def test_generator_batch(image, *, netG):
    netG.eval()
    output = netG(image)
    return output


@MODELS.register_module()
class BasicRestorer(BaseModel):
    """Basic model for image restoration.

    It must contain a generator that takes an image as inputs and outputs a
    restored image. It also has a pixel-wise loss for training.

    The subclasses should overwrite the function `forward_train`,
    `forward_test` and `train_step`.

    Args:
        generator (dict): Config for the generator structure.
        pixel_loss (dict): Config for pixel-wise loss.
        train_cfg (dict): Config for training. Default: None.
        eval_cfg (dict): Config for testing. Default: None.
        pretrained (str): Path for pretrained model. Default: None.
    """
    allowed_metrics = {'PSNR': psnr, 'SSIM': ssim}

    def __init__(self, generator, pixel_loss, train_cfg=None, eval_cfg=None, pretrained=None):
        super(BasicRestorer, self).__init__()

        self.train_cfg = train_cfg
        self.eval_cfg = eval_cfg

        # generator
        self.generator = build_backbone(generator)
        # loss
        self.pixel_loss = build_loss(pixel_loss)

        # load pretrained
        self.init_weights(pretrained)

        self.data = mge.tensor()
        self.label = mge.tensor()

    def init_weights(self, pretrained=None):
        """Init weights for models.

        Args:
            pretrained (str, optional): Path for pretrained weights. If given
                None, pretrained weights will not be loaded. Defaults to None.
        """
        self.generator.init_weights(pretrained)

    def train_step(self, batchdata):
        """train step.

        Args:
            batchdata: list for train_batch, numpy.ndarray, length up to Collect class.
        Returns:
            list: loss
        """
        data, label = batchdata
        self.optimizers['generator'].zero_grad()
        loss = train_generator_batch(data, label, opt=self.optimizers['generator'], netG=self.generator, netloss=self.pixel_loss)
        self.optimizers['generator'].step()
        return loss
```
Loss:
```

@LOSSES.register_module()
class L1Loss(M.Module):
    def __init__(self):
        super(L1Loss, self).__init__()

    def forward(self, pred, label):
        return mgeloss.l1_loss(pred = pred, label = label)
```
train_step是指对模型参数做一次迭代，传入的batchdata是一个list[ndarray]，长度为2，shape分别是[b,3,40,40]和[b,3,160,160]，代表低分辨率图像和高分辨率图像。
模型是一个经典的超分辨网络DBPN(https://arxiv.org/abs/1904.05677)。

**Comment by xxr3376** [comment]
@Feynman1999 看起来你的训练代码没什么问题，有一个猜想麻烦帮忙验证

有可能是因为默认的 kernel 选择的不够优，你可以尝试在代码开头添加自动 kernel 选择的代码
```
from megengine.functional.debug_param import set_conv_execution_strategy
    set_conv_execution_strategy("HEURISTIC_REPRODUCIBLE")
```

如果还不是的话，可能需要你提供一个可完整运行的最小复现，我们来研究一下为什么显存偏大。

**Comment by MegChai** [comment]
由于这个 Issue 长时间没有得到答复，我们决定先关闭它。如果有任何后续信息或新的疑问，欢迎重新打开～
